### PR TITLE
Fixed behavior for scroll views that have a nonzero top UIEdgeInset

### PR DESCRIPTION
**Pull request by Baddaboo**
This change takes care of a long-standing to-do by @mglidden where YLRefreshHeaderView would reset the top content inset of the scroll view to zero, rather than the the previous inset, after the refresh completes. Open to suggestions and feedback!

### CODE REVIEW

File: Classes/YLRefreshHeaderView/YLRefreshHeaderView.m [hunk 1]
```diff
@@ -31,26 +34,32 @@ - (void)setRefreshState:(YLRefreshHeaderViewState)refreshState {
...
-    contentInset.top = (refreshState == YLRefreshHeaderViewStateRefreshing) ? self.pullAmountToRefresh : 0;
-    strongScrollView.contentInset = contentInset;
-  };
-  void (^completionBlock)(BOOL) = ^(BOOL finished) {
-    if (self.refreshState == YLRefreshHeaderViewStateClosing) {
-      self.refreshState = YLRefreshHeaderViewStateNormal;
+    void (^animationBlock)() = ^{
+      UIScrollView *strongScrollView = self.scrollView;
+      UIEdgeInsets contentInset = strongScrollView.contentInset;
+      if (refreshState == YLRefreshHeaderViewStateRefreshing) {
```
Comment (nchourrout) [inline Classes/YLRefreshHeaderView/YLRefreshHeaderView.m]: Consider using a `switch` statement here.

File: Classes/YLRefreshHeaderView/YLRefreshHeaderView.m [hunk 2]
```diff
@@ -10,6 +10,9 @@
 #import "YLRefreshHeaderViewPrivate.h"
 #import "YLRefreshHeaderViewSubclass.h"
 
+@interface YLRefreshHeaderView ()
+@property (nonatomic) CGFloat previousTopInset;
+@end
```
Comment (nchourrout) [inline Classes/YLRefreshHeaderView/YLRefreshHeaderView.m]: Can you add an empty line between `@end` and `@implementation`?

File: Classes/YLRefreshHeaderView/YLRefreshHeaderView.m [hunk 3]
```diff
@@ -10,6 +10,9 @@
 #import "YLRefreshHeaderViewPrivate.h"
 #import "YLRefreshHeaderViewSubclass.h"
 
+@interface YLRefreshHeaderView ()
+@property (nonatomic) CGFloat previousTopInset;
```
Comment (nchourrout) [inline Classes/YLRefreshHeaderView/YLRefreshHeaderView.m]: Can you add `assign`?

**Comment by SSheldon** [comment]
@thibaudrobelain, could you take a look at this PR? When it's got your sign off I'm happy to merge it.

File: Classes/YLRefreshHeaderView/YLRefreshHeaderView.m [hunk 4]
```diff
@@ -31,26 +35,37 @@ - (void)setRefreshState:(YLRefreshHeaderViewState)refreshState {
...
-  void (^completionBlock)(BOOL) = ^(BOOL finished) {
-    if (self.refreshState == YLRefreshHeaderViewStateClosing) {
-      self.refreshState = YLRefreshHeaderViewStateNormal;
+    void (^animationBlock)() = ^{
+      UIScrollView *strongScrollView = self.scrollView;
+      UIEdgeInsets contentInset = strongScrollView.contentInset;
+      switch (refreshState) {
+        case YLRefreshHeaderViewStateRefreshing:
+          self.previousTopInset = contentInset.top;
+          contentInset.top = self.pullAmountToRefresh;
```
Comment (thibaudrobelain) [inline Classes/YLRefreshHeaderView/YLRefreshHeaderView.m]: You'll need to add self.previousTopInset here also

Comment (thibaudrobelain) [inline Classes/YLRefreshHeaderView/YLRefreshHeaderView.m]: You'll need to add self.previousTopInset here

File: Classes/YLRefreshHeaderView/YLRefreshHeaderView.m [hunk 5]
```diff
@@ -83,4 +98,4 @@ - (void)containingScrollViewDidEndDragging:(UIScrollView *)scrollView {
   }
```
Comment (thibaudrobelain) [inline Classes/YLRefreshHeaderView/YLRefreshHeaderView.m]: In here and in the method below, scrollView.contentOffset.y is used to determine whether we pulled far enough to refresh. However, if the scrollview has a contentInset > 0, scrollView.contentOffset will directly be -scrollView.contentInset.top. You thus need a helper method to give you the relativeContentOffset of the scrollview that will abstract the contentInset part of things, something like "self.scrollView.contentInset.top + self.scrollView.contentOffset.y";
You can then replace every place using the scrollView.contentOffset.y with this and it should work.

File: Classes/YLRefreshHeaderView/YLRefreshHeaderView.m (hunk 4 above)
Comment (thibaudrobelain) [inline Classes/YLRefreshHeaderView/YLRefreshHeaderView.m]: You'll need to add self.previousTopInset here

Comment (Baddaboo) [inline Classes/YLRefreshHeaderView/YLRefreshHeaderView.m]: Sorry, I'm not sure I understand this note: I use self.previousTopInset to hold the inset value in L46 before it's set to the refresh header height in L47. Is there another place I should be setting it, or using it?

Comment (thibaudrobelain) [inline Classes/YLRefreshHeaderView/YLRefreshHeaderView.m]: So, while trying out the branch I saw that if you have UIRectEdgeTop enabled, your contentInset.top should be something like 64. However, if my pullAmountToRefresh is 15 for instance, I would swap this 64 with 15 which would mess around the view. What I meant is that it probably needs 64 + the pullAmountToRefresh, so basically:
contentInset.top = self.previousTopInset + self.pullAmountToRefresh.

To test this correctly you can open the YLTableViewExample workspace and change UIRectEdgeNone to UIRectEdgeTop on L31 of YLExampleViewController. You'll see that if you use this branch as is it will not produce the expected behavior

Comment (Baddaboo) [inline Classes/YLRefreshHeaderView/YLRefreshHeaderView.m]: Ah, I'm guessing the 64px is the height of the navigation bar overlapping the tableview. I'm not sure if this is something I should change, because the previous implementation assumed that contentInset.top is initially zero, resets to zero, and that the top edge of the table view is aligned with the bottom edge of the navigation bar. So if you have a pullAmountToRefresh of 15, the new contentInset.top would be 15 as well. Implementations that use YLTableView may already rely on this behavior.  This would present a bigger challenge when contentInset.top is initially negative, which would mean part of the refresh header would be cut off

Comment (jimmynotjames) [inline Classes/YLRefreshHeaderView/YLRefreshHeaderView.m]: Blake, I don't understand this response. What about the example
implementation breaking? The bottom line is that the behavior is not the
correct one given the example. Can you fix it and still keep that example
code from breaking? Are you asking for help to find the solution?

--Jimmy Ho
iOS Engineer | Reader Experience Team

Comment (Baddaboo) [inline Classes/YLRefreshHeaderView/YLRefreshHeaderView.m]: Oh wait, I see what you mean now. This is a bit related to the previous comment about a positive contentInset.top. I think we can account for this case

File: Classes/YLRefreshHeaderView/YLRefreshHeaderView.m (hunk 5 above)
Comment (Baddaboo) [inline Classes/YLRefreshHeaderView/YLRefreshHeaderView.m]: Sounds good!

File: Classes/YLRefreshHeaderView/YLRefreshHeaderView.m [hunk 6]
```diff
@@ -74,24 +74,28 @@ - (CGFloat)pullAmountToRefresh {
   return 0;
 }
 
+- (CGFloat)relativeYContentOffsetForScrollView:(UIScrollView *)scrollView{
```
Comment (thibaudrobelain) [inline Classes/YLRefreshHeaderView/YLRefreshHeaderView.m]: nit: _ before "relative.." and space after scrollView

File: Classes/YLRefreshHeaderView/YLRefreshHeaderView.m [hunk 7]
```diff
@@ -74,24 +74,28 @@ - (CGFloat)pullAmountToRefresh {
...
+    if (self.refreshState == YLRefreshHeaderViewStateReadyToRefresh && [self relativeYContentOffsetForScrollView:scrollView] > -self.pullAmountToRefresh && [self relativeYContentOffsetForScrollView:scrollView] < 0) {
       self.refreshState = YLRefreshHeaderViewStateNormal;
     // If we've dragged far enough, put us in the Ready to Refresh state
-    } else if (self.refreshState == YLRefreshHeaderViewStateNormal && scrollView.contentOffset.y <= -self.pullAmountToRefresh) {
+    } else if (self.refreshState == YLRefreshHeaderViewStateNormal && [self relativeYContentOffsetForScrollView:scrollView] <= -self.pullAmountToRefresh) {
       self.refreshState = YLRefreshHeaderViewStateReadyToRefresh;
     }
   }
-  self.currentPullAmount = MAX(0, -scrollView.contentOffset.y);
+  self.currentPullAmount = MAX(0, -[self relativeYContentOffsetForScrollView:scrollView]);
```
Comment (thibaudrobelain) [inline Classes/YLRefreshHeaderView/YLRefreshHeaderView.m]: This breaks with a tableview that has a negative top inset

**Comment by thibaudrobelain** [comment]
Could you try to add tests in the test target that describe a few edge cases of working with contentInset, making sure that if you change the contentOffset by less < pullAmount or > pullAmount it updates the refreshState property properly? That way we can be sure that your changes don't break a specific weird edge case.

**Comment by Baddaboo** [comment]
@thibaudrobelain I've added UI tests to the example project, but because the test target is isolated from the app target, I can't have the test directly modify the content inset of the app for each case. Also, since the refresh header requires simulated gestures to trigger, a unit test wouldn't be too helpful. Hopefully this should cover the uses we want :)

**Comment by thibaudrobelain** [comment]
@Baddaboo, just checking in on how it's going, are you going to try and wrap this up?

**Comment by SSheldon** [comment]
I'm assuming we've decided to abandon this PR, sounds like there were still issues to resolve and we didn't end up needing it? Feel free to reopen otherwise.
